lib/routes/account.js

```diff
@@ -973,78 +977,93 @@ module.exports = (log, db, mailer, Password, config, customs, signinUtils, push)
         }
         request.setMetricsFlowCompleteSignal(flowCompleteSignal)
 
-        return resetAccountData()
+        return checkRecoveryKey()
+          .then(resetAccountData)
+          .then(deleteRecoveryKey)
           .then(createSessionToken)
           .then(createKeyFetchToken)
           .then(recordSecurityEvent)
           .then(createResponse)
           .then(reply, reply)
 
+        function checkRecoveryKey() {
+          if (recoveryKeyId && wrapKb) {
+            return db.getRecoveryKey(accountResetToken.uid, recoveryKeyId)
+          }
+
+          if (recoveryKeyId && ! wrapKb || wrapKb && ! recoveryKeyId) {
+            throw error.invalidRequestParameter()
```

Comment:
You can move this into the route validation using Joi's `.and()` feature, see e.g:

  https://github.com/mozilla/fxa-auth-server/blob/master/lib/routes/devices-and-sessions.js#L141

In this case I think we'd want `.and('wrapkB', 'recoveryKeyId')` in the route validator definition.

Reply:
This is exactly what I was looking for!